backend/controllers/saleController.js

```diff
@@ -0,0 +1,127 @@
+const Dealership = require('../models/Dealership');
+const User = require('../models/User');
+const Sale = require('../models/VehicleSale');
+const Vehicle = require('../models/Vehicle');
+
+const ErrorResponse = require('../utils/errorResponse');
+const asyncHandler = require('../middleware/async');
+const advancedFilter = require('../utils/advancedFilter');
+
+// @desc    Create a sale instance
+// @route   POST /api/v1/inventory/details/sale'
+// @access  Public
+exports.createSale = asyncHandler(async (req, res, next) => {
+    //Check if there is a sale object for a given vehicle
+    const sale_check = await Sale.findOne({vehicle: req.body.vehicle});
+
+    // if a sale has already been created for this object return an error
+    if (!!sale_check) {
+        return next(
+            new ErrorResponse(`A sale has already been created for this vehicle`, 400)
+        );
+    }
+    // create new sale object with the data passed in the request body
+    const sale = await Sale.create(req.body);
+    //Map this created sale to the vehicle
+    const newVehicle = await Vehicle.findByIdAndUpdate(req.body.vehicle, {sale: sale._id}, {
+        runValidators: true,
+        new: true
+    });
+    // send response
+    res.status(201).json({
+        success: true,
+        payload: sale
+    });
+});
+
+// @desc    Get a sale instance
+// @route   GET /api/v1/inventory/details/sale/:dealershipId'
+// @access  Public
+exports.getSales = asyncHandler(async (req, res, next) => {    
+    // grab the dealership_ID from the body and verify that the dealership exists
+    const dealership = await Dealership.findById(req.params.dealershipId);
+    
+    // no dealership found
+    if (!dealership) {
+        return next(
+            new ErrorResponse(`This dealership ID ${req.params.dealershipId} with this not found. Cannot return a list of vehicls without a valid dealership.`, 404)
+        );
+    }
+
+    const sales = await Sale.find({ dealership: req.params.dealershipId })
+    .populate('vehicle')
+    .populate('sales_rep')
+    .populate('approved_by');
```

Comment:
Instead of populating manually the different properties, you could take advantage of the `advancedFilter` util function.

Reply:
I'm not sure how to leverage the advancedFilter fucntion. Could you provide an example?

Reply:
You could use something like this 
`const query = await advancedFilter(Sale, req.query)`

However, adding this change means you need to modify your API request (axios) by adding the populate argument as a query parameter like so:
`GET /api/v1/inventory/details/sale/:dealershipId/?populate=vehicle,sales_rep,approved_by`.

This makes the controller more flexible in the future. If not, the lines of code could also be reduced by simply using `.populate('vehicle sales_rep approved_by')`

Reply:
Thanks for the example. I have opted for the second option (i.e. populate separated by space). This is address in this [commit](https://github.com/purplesmurf1998/Autotracks/pull/52/commits/6d067bd2b87c4a5d34eede74a10719c101e6f202)